src/main/java/com/troblecodings/signals/handler/NameHandler.java

```diff
@@ -177,79 +169,103 @@ private static void createToFile(final NameStateInfo info, final String name) {
-                final String name = ALL_LEVEL_FILES.get(world).getString(pos);
-                if (name.isEmpty())
-                    return;
-                synchronized (ALL_NAMES) {
-                    ALL_NAMES.put(info, name);
-                }
-                states.add(info);
-                sendNameToClient(info, name);
+                states.add(new NameStateInfo(world, pos));
             }
         });
-        synchronized (CURRENTLY_LOADED_CHUNKS) {
-            CURRENTLY_LOADED_CHUNKS.put(chunk, states);
-        }
+        loadNames(states, player);
     }
 
     @SubscribeEvent
-    public static void onChunkUnload(final ChunkEvent.Unload event) {
-        final ChunkAccess chunk = event.getChunk();
-        final Level level = (Level) chunk.getWorldForge();
-        if (level.isClientSide())
+    public static void onChunkUnWatch(final ChunkWatchEvent.UnWatch event) {
+        final ServerLevel world = event.getWorld();
+        if (world.isClientSide)
             return;
-        List<NameStateInfo> states;
-        synchronized (CURRENTLY_LOADED_CHUNKS) {
-            states = CURRENTLY_LOADED_CHUNKS.remove(chunk);
-        }
-        states.forEach(stateInfo -> {
-            String name;
-            synchronized (ALL_NAMES) {
-                name = ALL_NAMES.remove(stateInfo);
+        final ChunkAccess chunk = world.getChunk(event.getPos().getWorldPosition());
+        final List<NameStateInfo> states = new ArrayList<>();
+        chunk.getBlockEntitiesPos().forEach(pos -> {
+            final Block block = chunk.getBlockState(pos).getBlock();
+            if (block instanceof Signal || block instanceof RedstoneIO) {
+                states.add(new NameStateInfo(world, pos));
             }
-            if (name == null)
-                return;
-            createToFile(stateInfo, name);
         });
+        unloadNames(states);
     }
 
-    @SubscribeEvent
-    public static void onPlayerJoin(final PlayerEvent.PlayerLoggedInEvent event) {
-        final Player player = event.getPlayer();
-        final Map<NameStateInfo, String> names;
-        synchronized (ALL_NAMES) {
-            names = ImmutableMap.copyOf(ALL_NAMES);
-        }
-        names.forEach((info, name) -> sendTo(player, packToBuffer(info.pos, name)));
+    private static void loadNames(final List<NameStateInfo> infos, final @Nullable Player player) {
+        if (infos == null || infos.isEmpty())
+            return;
+        new Thread(() -> {
+            infos.forEach(info -> {
+                synchronized (LOAD_COUNTER) {
+                    Integer count = LOAD_COUNTER.get(info);
+                    if (count != null && count > 0) {
+                        LOAD_COUNTER.put(info, ++count);
+                        return;
+                    }
+                    LOAD_COUNTER.put(info, 1);
+                    final String name = ALL_LEVEL_FILES.get(info.world).getString(info.pos);
```

Comment:
ALL_LEVEL_FILES muss synchronisiert werden da es eine critical section ist
```suggestion
                    String name;
                    synchronized(ALL_LEVEL_FILES) {
                       name = ALL_LEVEL_FILES.get(info.world).getString(info.pos);
                    }
```